Add longer length support to .ahk

onerandomusername opened the pull request:
If responses are too long, upload to p.ahkscript,org.
If its longer than that, upload as a file.
Also added support to run code from p.ahkscript.org

onerandomusername committed:
Add longer length support to .ahk

If responses are too long, upload to p.ahkscript,org.
If its longer than that, upload as a file.
Also added support to run code from p.ahkscript.org

onerandomusername committed:
Do not use the pastebin if not the AHK guild

Also lengthen the limit to support showing errors.

Run1e commented on cogs/ahk/ahk.py:
```diff
@@ -5,6 +5,7 @@
 from collections import OrderedDict
 from datetime import datetime, timedelta, timezone
 
+import io
```
Standard library import belongs further up in the imports

Run1e commented on cogs/ahk/ahk.py:
```diff
@@ -295,24 +296,50 @@ async def cloudahk_call(self, ctx, code, lang='ahk'):
 
 		stdout = stdout.replace('`', '`\u200b')
+		file = None
+		encoded_out = stdout.encode('utf-8')
```
Is stdout not already unicode?

onerandomusername replied:
I don't believe it is.

Run1e commented on cogs/ahk/ahk.py:
```diff
@@ -295,24 +296,50 @@ async def cloudahk_call(self, ctx, code, lang='ahk'):
 		stdout = stdout.replace('`', '`\u200b')
+		file = None
+		encoded_out = stdout.encode('utf-8')
+		if len(stdout) < 1800 and stdout.count('\n') < 20 and stdout.count('\r') < 20:
```
1. Why is `len(stdout)` used here instead of `len(encoded_out)`? This can be put into a variable before the if/else ladder.
2. Why are line feeds and carriage returns counted seperately here? Carriage returns can be removed from the string before processing it to avoid this.

Run1e commented on cogs/ahk/ahk.py:
````diff
@@ -295,24 +296,50 @@ async def cloudahk_call(self, ctx, code, lang='ahk'):
+		encoded_out = stdout.encode('utf-8')
+		if len(stdout) < 1800 and stdout.count('\n') < 20 and stdout.count('\r') < 20:
+			#upload as plaintext
+			valid_response = 'No output.\n' if stdout == '' else '\n```autoit\n{0}\n```'.format(stdout)
````
Repeated `valid_response` assignments. Practice dry code

Run1e commented on cogs/ahk/ahk.py:
````diff
@@ -295,24 +296,50 @@ async def cloudahk_call(self, ctx, code, lang='ahk'):
+		if len(stdout) < 1800 and stdout.count('\n') < 20 and stdout.count('\r') < 20:
+			#upload as plaintext
+			valid_response = 'No output.\n' if stdout == '' else '\n```autoit\n{0}\n```'.format(stdout)
+		elif len(encoded_out) < 100000 and AHK_GUILD_ID == ctx.guild.id:
````
What's the significance of 100000?

onerandomusername replied:
A limit of 100kb to the pastebin.
```suggestion
		elif len(encoded_out) < 100000 and AHK_GUILD_ID == ctx.guild.id: # 100kb
```

Run1e commented on cogs/ahk/ahk.py:
````diff
@@ -295,24 +296,50 @@ async def cloudahk_call(self, ctx, code, lang='ahk'):
+			valid_response = 'No output.\n' if stdout == '' else '\n```autoit\n{0}\n```'.format(stdout)
+		elif len(encoded_out) < 100000 and AHK_GUILD_ID == ctx.guild.id:
+			#upload to ahkscript.org
+			data = {'name': 'Ace Bot (Clone)', 'code': stdout}
````
This is not an AceBot clone.

Run1e commented on cogs/ahk/ahk.py:
```diff
@@ -295,24 +296,50 @@ async def cloudahk_call(self, ctx, code, lang='ahk'):
+		elif len(encoded_out) < 100000 and AHK_GUILD_ID == ctx.guild.id:
+			#upload to ahkscript.org
+			data = {'name': 'Ace Bot (Clone)', 'code': stdout}
+			async with ctx.http.post('https://p.ahkscript.org/',data=data) as resp:
```
Missing space after comma

Run1e commented on cogs/ahk/ahk.py:
```diff
@@ -295,24 +296,50 @@ async def cloudahk_call(self, ctx, code, lang='ahk'):
+				if resp.status != 200:
+					raise commands.CommandError('Pastebin failed.')
+				valid_response = ' No output.\n' if stdout == '' else f' Results too large. Pasted. {resp.url}\n'
+		elif len(encoded_out) < (1000*1000*4000): # should be 4mb
```
`1000*1000*4000` can be put into a constant, and also missing spaces around *

Also max upload size for bots (and non-nitro users) is 8MB not 4MB

Run1e commented on cogs/ahk/ahk.py:
```diff
@@ -295,24 +296,50 @@ async def cloudahk_call(self, ctx, code, lang='ahk'):
+					raise commands.CommandError('Pastebin failed.')
+				valid_response = ' No output.\n' if stdout == '' else f' Results too large. Pasted. {resp.url}\n'
+		elif len(encoded_out) < (1000*1000*4000): # should be 4mb
+			fp = io.BytesIO(stdout.encode('utf-8'))
```
Invoking .encode() twice

Run1e commented on cogs/ahk/ahk.py:
```diff
@@ -295,24 +296,50 @@ async def cloudahk_call(self, ctx, code, lang='ahk'):
+		try:
+			await ctx.reply(content=out, file=file)
+		except discord.HTTPException:
+			await ctx.send(content=out, file=file)
```
If .reply fails then .send will (likely) fail - this logic doesn't really make sense

onerandomusername replied:
Replying will fail if the initial message was deleted, this is to silently catch that error.

Run1e commented on cogs/ahk/ahk.py:
```diff
@@ -295,24 +296,50 @@ async def cloudahk_call(self, ctx, code, lang='ahk'):
+					code = await resp.text()
+
 
 		stdout, time = await self.cloudahk_call(ctx, code)
```
This will be called even if the API call above fails.

onerandomusername replied:
That _was_ intentional because if its not a valid link, assume that the user wanted to run that as code.

Run1e reviewed (changes requested):
Why do we need to be able to upload to both discord and the pastebin? I would argue it makes most sense to upload as text if under 1800 characters, and if below 8MB upload to discord.

onerandomusername committed:
Made suggested changes

Removed dry code.
Used the encoded_stdout where it made sense.
Made other suggested changes.

onerandomusername committed:
Adding replies requires discord.py v1.6.0

onerandomusername committed:
Remove pasting to the pastebin & update endpoint support

onerandomusername committed:
Sorted Imports

onerandomusername commented:
This is ready for review again.

onerandomusername committed:
Make suggested changes.

onerandomusername committed:
improve grave accent escapes

Changed escaping grave accents (\`) to only escape if it would end the current block, and only if its plaintext.
There's nothing worse than a huge file that has every single \` with an invisible character following it...

onerandomusername committed:
Fix grave accent escape not always escaping
